File: frontend/src/types/connector.ts
export interface ConnectorActionResponse {
  id: string
  connector_id: string
  name: string
  description: string | null
  method: string // GET | POST | PUT | DELETE
  path: string
  parameters_schema: Record<string, unknown> | null
  request_body_template: Record<string, unknown> | null
  response_extract: string | null
  requires_confirmation: boolean
  created_at: string
  updated_at: string | null
}

export interface ConnectorActionCreate {
  name: string
  description?: string | null
  method?: string
  path: string
  parameters_schema?: Record<string, unknown> | null
  request_body_template?: Record<string, unknown> | null
  response_extract?: string | null
  requires_confirmation?: boolean
}

export interface ConnectorActionUpdate {
  name?: string
  description?: string | null
  method?: string
  path?: string
  parameters_schema?: Record<string, unknown> | null
  request_body_template?: Record<string, unknown> | null
  response_extract?: string | null
  requires_confirmation?: boolean
}

// Database connection config
export interface DbConnectionConfig {
  driver: string // "postgresql" | "mysql" | "oracle" | "sqlserver" | "dm8" | "kingbasees" | "gbase" | "highgo"
  host: string
  port: number
  database: string
  schema?: string // PG/Oracle only
  username: string
  password?: string // Only in create/update, masked in response
  encrypted_password?: string // In response, always masked
  ssl: boolean
  ca_cert?: string
  read_only: boolean
  max_rows: number
  query_timeout: number
}

export interface ConnectorResponse {
  id: string
  name: string
  description: string | null
  icon: string | null
  type: string // "api" | "database"
  base_url: string | null
  auth_type: string // "api_key" | "bearer" | "oauth2" | "basic" | "none"
  auth_config: Record<string, unknown> | null
  db_config?: DbConnectionConfig | null // Present when type="database"
  is_official: boolean
  is_active: boolean
  forked_from: string | null
  version: number
  visibility: string // "personal" | "org" | "global"
  org_id: string | null
  user_id: string // connector owner id
  has_default_credentials: boolean
  allow_fallback: boolean
  publish_status?: string | null // "pending_review" | "approved" | "rejected"
  reviewed_by?: string | null
  reviewed_at?: string | null
  review_note?: string | null
  actions: ConnectorActionResponse[]
  created_at: string
  updated_at: string | null
}

export interface ConnectorCreate {
  name: string
  description?: string | null
  icon?: string | null
  type?: string
  base_url: string
  auth_type?: string
  auth_config?: Record<string, unknown> | null
}

export interface ConnectorUpdate {
  name?: string
  description?: string | null
  icon?: string | null
  type?: string
  base_url?: string
  auth_type?: string
  auth_config?: Record<string, unknown> | null
  db_config?: DbConnectionConfig | null
  allow_fallback?: boolean
}

export interface CredentialUpsertRequest {
  token?: string | null
  api_key?: string | null
  username?: string | null
  password?: string | null
}

export interface MyCredentialStatus {
  has_credentials: boolean
  auth_type: string
  allow_fallback: boolean
}

export interface OpenAPIImportRequest {
  spec?: Record<string, unknown>
  spec_url?: string
  spec_raw?: string
  replace_existing?: boolean
}

export interface AIGenerateActionsRequest {
  instruction: string
  context?: string
}

export interface AIRefineActionRequest {
  instruction: string
  action_id?: string
}

export interface AIActionResult {
  created: ConnectorActionResponse[]
  updated: ConnectorActionResponse[]
  deleted: string[]
  connector_updated: ConnectorResponse | null
  message: string
  message_key?: string
  message_args?: Record<string, unknown>
}

export interface AICreateConnectorResult {
  connector: ConnectorResponse
  message: string
  message_key?: string
  message_args?: Record<string, unknown>
}

// Database connector create
export interface DbConnectorCreate {
  name: string
  description?: string | null
  icon?: string | null
  type: "database"
  db_config: DbConnectionConfig
}

// Schema types
export interface SchemaTable {
  id: string
  table_name: string
  display_name: string | null
  description: string | null
  is_visible: boolean
  columns: SchemaColumn[]
}

export interface SchemaColumn {
  id: string
  column_name: string
  display_name: string | null
  description: string | null
  data_type: string
  is_nullable: boolean
  is_primary_key: boolean
  is_visible: boolean
}

export interface SchemaTableUpdate {
  display_name?: string
  description?: string
  is_visible?: boolean
}

export interface SchemaColumnUpdate {
  display_name?: string
  description?: string
  is_visible?: boolean
}

// Test connection
export interface TestConnectionResponse {
  success: boolean
  db_version: string | null
  error: string | null
}

// Introspect response
export interface IntrospectResponse {
  tables_discovered: number
  columns_discovered: number
}

// Query
export interface QueryRequest {
  sql: string
}

export interface QueryResponse {
  columns: string[]
  rows: unknown[][]
  row_count: number
  truncated: boolean
  execution_time_ms: number
  error: string | null
}

// AI annotate
export interface AIAnnotateResponse {
  annotated_count: number
  preview: Record<string, unknown>[]
}

export interface AIAnnotateJobStarted {
  job_id: string
  table_count: number
}

export interface AIAnnotateJobStatus {
  job_id: string
  status: "pending" | "running" | "done" | "error"
  completed_batches: number
  total_batches: number
  annotated_count: number
  error: string | null
}

// Connector template (built-in, API-only)
export interface ConnectorTemplate {
  id: string
  name: string
  description: string
  icon?: string | null
  category: string
  blueprint: Record<string, unknown>
}

File: src/fim_agent/web/schemas/connector.py
"""Connector request/response schemas."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


# --- Action Schemas ---


class ActionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    method: str = "GET"
    path: str = Field(min_length=1, max_length=500)
    parameters_schema: dict[str, Any] | None = None
    request_body_template: dict[str, Any] | None = None
    response_extract: str | None = None
    requires_confirmation: bool = False


class ActionUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    method: str | None = None
    path: str | None = None
    parameters_schema: dict[str, Any] | None = None
    request_body_template: dict[str, Any] | None = None
    response_extract: str | None = None
    requires_confirmation: bool | None = None


class ActionResponse(BaseModel):
    id: str
    connector_id: str
    name: str
    description: str | None
    method: str
    path: str
    parameters_schema: dict[str, Any] | None
    request_body_template: dict[str, Any] | None
    response_extract: str | None
    requires_confirmation: bool
    created_at: str
    updated_at: str | None


# --- OpenAPI Import ---


class OpenAPIImportRequest(BaseModel):
    """Accepts an OpenAPI spec via one of three input modes."""

    spec: dict[str, Any] | None = None
    spec_url: str | None = None
    spec_raw: str | None = None
    replace_existing: bool = False


# --- Connector Schemas ---


class ConnectorCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    icon: str | None = None
    type: str = "api"
    base_url: str = Field(min_length=1, max_length=500)
    auth_type: str = "none"
    auth_config: dict[str, Any] | None = None


class ConnectorUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    icon: str | None = None
    type: str | None = None
    base_url: str | None = None
    auth_type: str | None = None
    auth_config: dict[str, Any] | None = None


class ConnectorResponse(BaseModel):
    id: str
    name: str
    description: str | None
    icon: str | None
    type: str
    base_url: str
    auth_type: str
    auth_config: dict[str, Any] | None
    is_official: bool
    forked_from: str | None
    version: int
    actions: list[ActionResponse]
    created_at: str
    updated_at: str | None


# --- AI Action Schemas ---


class AIGenerateActionsRequest(BaseModel):
    instruction: str = Field(min_length=1, max_length=2000)
    context: str | None = Field(default=None, max_length=10000)


class AIRefineActionRequest(BaseModel):
    instruction: str = Field(min_length=1, max_length=2000)
    action_id: str | None = None


class AIActionResult(BaseModel):
    created: list[ActionResponse] = []
    updated: list[ActionResponse] = []
    deleted: list[str] = []
    failed: list[str] = []
    connector_updated: ConnectorResponse | None = None
    message: str = ""
    message_key: str = ""
    message_args: dict = {}


class AICreateConnectorRequest(BaseModel):
    instruction: str = Field(min_length=1, max_length=5000)


class AICreateConnectorResult(BaseModel):
    connector: ConnectorResponse
    message: str = ""
    message_key: str = ""
    message_args: dict = {}
